File: C#/laborator3CSharpGUI/lab3/lab3/repo/OficiuRepository.cs
using lab3.domain;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using log4net;
using System.Data;

namespace lab3.repo
{
    public class OficiuRepository : ICrudRepository<string, Oficiu>
    {
        private static readonly ILog log = LogManager.GetLogger("OficiuRepository");
        public OficiuRepository(){ log.Info("Initializing OficiuRepository"); }

        public void delete(string id)
        {
            log.InfoFormat("Deleting entity with id {0}", id);
            IDbConnection connection = DBUtils.getConnection();

            using (var command  = connection.CreateCommand())
            {
                command.CommandText = "delete from Oficiu where idOficiu=@id";

                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No entity deleted!");
            }
        }

        public IEnumerable<Oficiu> findAll()
        {
            IDbConnection connection = DBUtils.getConnection();
            IList<Oficiu> oficiuList = new List<Oficiu>();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select idOficiu, username, password from Oficiu";

                using (var result = command.ExecuteReader())
                {
                    while (result.Read())
                    {
                        int id = result.GetInt32(0);
                        string username = result.GetString(1);
                        string password = result.GetString(2);
                        Oficiu oficiu = new Oficiu(username, password)
                        {
                            Id = id.ToString()
                        };
                        oficiuList.Add(oficiu);
                    }
                }
            }

            return oficiuList;
        }

        public Oficiu findOne(string id)
        {
            log.InfoFormat("Finding entity with id {0}", id);
            IDbConnection connection = DBUtils.getConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select username, password from Oficiu where idOficiu=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);

                using (var result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        string username = result.GetString(0);
                        string password = result.GetString(1);
                        Oficiu oficiu = new Oficiu(username, password)
                        {
                            Id = id
                        };
                        log.InfoFormat("Exiting with entity found {0}", oficiu);
                        return oficiu;
                    }
                }
            }

            log.InfoFormat("No entity with id {0} found", id);
            return null;
        }

        public Oficiu findOne(string user, string pass)
        {
            log.InfoFormat("Finding entity with username {0} and password {1}", user, pass);
            IDbConnection connection = DBUtils.getConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText =
                    "select idOficiu, username, password from Oficiu where username=@u and password=@p";
                IDbDataParameter paramUser = command.CreateParameter();
                paramUser.ParameterName = "@u";
                paramUser.Value = user;
                command.Parameters.Add(paramUser);

                IDbDataParameter paramPass = command.CreateParameter();
                paramPass.ParameterName = "@p";
                paramPass.Value = pass;
                command.Parameters.Add(paramPass);

                using (var result = command.ExecuteReader())
                {
                    if (result.Read())
                    {
                        string idOficiu = result.GetInt32(0).ToString();
                        string username = result.GetString(1);
                        string password = result.GetString(2);
                        Oficiu oficiu = new Oficiu(username, password)
                        {
                            Id = idOficiu
                        };
                        log.InfoFormat("Exiting with entity found {0}", oficiu);
                        return oficiu;
                    }
                }
            }

            log.InfoFormat("No entity with username {0} and password {1}", user, pass);
            return null;
        }

        public void save(Oficiu entity)
        {
            log.InfoFormat("Saving entity {0}", entity);
            IDbConnection connection = DBUtils.getConnection();

            using (var command = connection.CreateCommand())
            {
                command.CommandText = "insert into Oficiu values (@idOficiu, @username, @password)";

                var paramId = command.CreateParameter();
                paramId.ParameterName = "@idOficiu";
                paramId.Value = entity.Id;
                command.Parameters.Add(paramId);

                var paramUser = command.CreateParameter();
                paramUser.ParameterName = "@username";
                paramUser.Value = entity.Username;
                command.Parameters.Add(paramUser);

                var paramPass = command.CreateParameter();
                paramPass.ParameterName = "@password";
                paramPass.Value = entity.Password;
                command.Parameters.Add(paramPass);

                var result = command.ExecuteNonQuery();
                if (result == 0)
                    throw new RepositoryException("No entity added!");
            }
        }
    }
}

File: Java/laborator3JavaGUI/src/main/java/repo/OficiuRepository.java
package repo;

import domain.Oficiu;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class OficiuRepository implements IOficiuRepository{
    private static final Logger logger = LogManager.getLogger(OficiuRepository.class);
    private JdbcUtils jdbcUtils;

    public OficiuRepository(Properties properties) {
        logger.info("Initializing OficiuRepository with properties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public void save(Oficiu entity) {
        logger.traceEntry("Saving entity {}", entity);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("insert into Oficiu values (?,?,?)")){
            preparedStatement.setString(1, entity.getId());
            preparedStatement.setString(2, entity.getUsername());
            preparedStatement.setString(3, entity.getPassword());
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit();
    }

    @Override
    public void update(String s, Oficiu entity) {
        logger.traceEntry("Updating entity with id {}", s);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("update Oficiu set username=?, password=? where idOficiu=?")){
            preparedStatement.setString(1, entity.getUsername());
            preparedStatement.setString(2, entity.getPassword());
            preparedStatement.setString(3, s);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(String s) {
        logger.traceEntry("Deleting entity with id {}", s);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("delete from Oficiu where idOficiu=?")){
            preparedStatement.setString(1, s);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit();
    }

    @Override
    public Oficiu findOne(String s) {
        logger.traceEntry("Finding entity with id {}", s);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from Oficiu where idOficiu=?")){
            preparedStatement.setString(1, s);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    String id = resultSet.getString("idOficiu");
                    String username = resultSet.getString("username");
                    String password = resultSet.getString("password");
                    Oficiu oficiu = new Oficiu(username, password);
                    oficiu.setId(id);
                    logger.traceExit(oficiu);
                    return oficiu;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit("No entity with id {} found", s);
        return null;
    }

    public Oficiu findByUserPass(String username, String password){
        logger.traceEntry("Finding entity with username {} and password {}", username, password);
        Connection connection = jdbcUtils.getConnection();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from Oficiu where username=? and password=?")){
            preparedStatement.setString(1, username);
            preparedStatement.setString(2, password);
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    String id = resultSet.getString("idOficiu");
                    String user = resultSet.getString("username");
                    String pass = resultSet.getString("password");
                    Oficiu oficiu = new Oficiu(user, pass);
                    oficiu.setId(id);
                    logger.traceExit(oficiu);
                    return oficiu;
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit("No entity with username " + username + " and password " + password);
        return null;
    }

    @Override
    public Iterable<Oficiu> findAll() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();
        List<Oficiu> oficiuList = new ArrayList<>();

        try(PreparedStatement preparedStatement = connection.prepareStatement("select * from Oficiu")){
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                while (resultSet.next()){
                    String id = resultSet.getString("idOficiu");
                    String username = resultSet.getString("username");
                    String password = resultSet.getString("password");
                    Oficiu oficiu = new Oficiu(username, password);
                    oficiu.setId(id);
                    oficiuList.add(oficiu);
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        logger.traceExit(oficiuList);
        return oficiuList;
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection connection = jdbcUtils.getConnection();

        try (PreparedStatement preparedStatement = connection.prepareStatement("select count(*) as [size] from Oficiu")){
            try(ResultSet resultSet = preparedStatement.executeQuery()){
                if(resultSet.next()){
                    logger.traceExit(resultSet.getInt("size"));
                    return resultSet.getInt("size");
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("database error..." + e);
        }
        return 0;
    }
}
